src/coord/src/coord.rs

```diff
@@ -2706,29 +2706,36 @@ impl Coordinator {
         // a larger timestamp and block, perhaps the user should intervene).
         let (index_ids, unmaterialized_source_ids) = self.catalog.nearest_indexes(uses_ids);
 
-        if !unmaterialized_source_ids.is_empty() {
-            coord_bail!(
-                "Unable to automatically determine a timestamp for your query; \
-                this can happen if your query depends on non-materialized sources.\n\
-                For more details, see https://materialize.com/s/non-materialized-error"
-            );
-        }
-
         // Determine the valid lower bound of times that can produce correct outputs.
         // This bound is determined by the arrangements contributing to the query,
         // and does not depend on the transitive sources.
-        let since = self.indexes.least_valid_since(index_ids.iter().cloned());
+        let mut since = self.indexes.least_valid_since(index_ids.iter().cloned());
```

Comment:
What do you think about the idea of doing the  join with `self.sources.least_valid_since` here? It feels like, if we have sources and they have since constraints, we want their constraints applied in all cases?

Reply:
I guess this would have some potentially new behavior that we would want to check (e.g. allowing AS OF-free queries against unmaterialized sources, and doing something perhaps surprising on sources without since reports (idk what non-durable sources look like)).

Much of this code is from an era where "source" is a proxy for "doesn't have `since` information", and if that has changed to something where "sources may have `since` information, or they may be volatile and not a great source of timestamp information" we might want to re-consider parts of it.

Reply:
I thought about doing the join here, but thought that would be confusing to readers of this function if we still disallowed select without as of on unmaterialized sources. Maybe: we should join with sources here after we answer all the questions from your second comment.